### ISSUE TITLE
Trouble understanding interface

### ISSUE THREAD

**Issue by AaronLiu1997** [HazyResearch/flash-attention#37]
In https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attention.py#L32, the comment says

```
            qkv: The tensor containing the query, key, and value. (B, S, 3, H, D) if key_padding_mask is None
                if unpadded: (nnz, 3, h, d)
            key_padding_mask: a bool tensor of shape (B, S)

```

Is it possible to call this function with a tensor of shape  (B, S, 3, H, D)  which *IS* padded? If so, how should the arguments be structured? Thanks!

**Comment by tridao** [comment]
Yes, it would then call [`unpad_input`](https://github.com/HazyResearch/flash-attention/blob/19d126102542086ba5dbaee07878220a24c58581/flash_attn/flash_attention.py#L56), which removes the padding tokens, then call flash-attention for unpadded inputs.

This pytorch module is meant to be a light wrapper around the [interface](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_interface.py), which then call into the CUDA functions.

**Comment by AaronLiu1997** [comment]
Thanks, I don't get the expected results when using this Pytorch module. Consider the following code:


```
import torch 
from flash_attn.flash_attention import FlashAttention


def flash_attn(qkv, bs, sq, num_heads, sm_scale, mask=None):
    attn = FlashAttention(softmax_scale = sm_scale)
    out = attn(qkv, key_padding_mask=mask)[0]
    return out.view(bs, sq, -1)

def std_attn_forward(qkv, B, N, num_heads, scale, mask=None):
    qkv = qkv.permute(2, 0, 3, 1, 4) # (3, bs, nhead, sq, hdim)
    q, k, v = qkv[0], qkv[1], qkv[2]   # (bs, nhead, sq, hdim))
    q = q * scale
    attn = (q @ k.transpose(-2, -1)) # (bs, nheads, sq-q, sq-k)
    if mask is not None:
        mask = mask.bool()
        attn = attn.masked_fill(~mask[:, None, None, :], float("-inf"))
    attn = attn.softmax(dim=-1).type_as(q)
    x = (attn @ v)    
    x = x.transpose(1, 2).reshape(B, N, -1)
    return x

if __name__ == '__main__':
    bs, sq, fdim, nheads, scale = 16, 256, 1024, 16, 0.1
    qkv = torch.randn(bs, sq, fdim*3).half().cuda()
    qkv = qkv.view(bs, sq, 3, nheads, -1)

    mask = None
    out1 = flash_attn(qkv, bs, sq, nheads, scale, mask=mask)
    out2 = std_attn_forward(qkv, bs, sq, nheads, scale, mask=mask)
    print('same with mask off:', torch.allclose(out1, out2, atol=1e-3))

    mask = (torch.arange(sq) < sq//2).unsqueeze(0).repeat(bs, 1).cuda()
    out1 = flash_attn(qkv, bs, sq, nheads, scale, mask=mask)
    out2 = std_attn_forward(qkv, bs, sq, nheads, scale, mask=mask)
    print('same with mask on:', torch.allclose(out1, out2, atol=1e-3))
```

From this I get

```
same with mask off: True
same with mask on: False
```

Is this how the pytorch module should be used with key masks? Thanks!

**Comment by tridao** [comment]
This is meant for self-attention so when you apply the key mask it also masks out the query.
It means that, for this case, the first 128 rows of the output from FlashAttention (`out1[:, :128]`) matches the first 128 rows of the reference (`out2[:, :128]`), while the remaining 128 rows are set to zero (you can print it out to check).

If you want to do cross attention (where q and k don't have the same length, and they could have different masks), then you would want to call [flash_attn_unpadded_func](https://github.com/HazyResearch/flash-attention/blob/19d126102542086ba5dbaee07878220a24c58581/flash_attn/flash_attn_interface.py#L213).

**Comment by AaronLiu1997** [comment]
Ah, got it, thanks for the clarification!!

I am however still having issues even after this fix, consider the following code:

```
import torch
from flash_attn.flash_attention import FlashAttention


def flash_attn(qkv, bs, sq, num_heads, sm_scale, mask=None):
    attn = FlashAttention(softmax_scale = sm_scale)
    out = attn(qkv, key_padding_mask=mask)[0]
    return out.view(bs, sq, -1)

def std_attn_forward(qkv, B, N, num_heads, scale, mask=None):
    qkv = qkv.permute(2, 0, 3, 1, 4) # (3, bs, nhead, sq, hdim)
    q, k, v = qkv[0], qkv[1], qkv[2]   # (bs, nhead, sq, hdim))
    q = q * scale
    attn = (q @ k.transpose(-2, -1)) # (bs, nheads, sq-q, sq-k)
    if mask is not None:
        mask = mask.bool()
        attn = attn.masked_fill(~mask[:, None, None, :], float("-inf"))
    attn = attn.softmax(dim=-1).type_as(q)
    x = (attn @ v)
    x = x.transpose(1, 2).reshape(B, N, -1)
    return x

if __name__ == '__main__':


    for nheads in [8, 16]:

        bs, sq, fdim, scale = 8, 256, 1024, 0.1
        qkv = torch.randn(bs, sq, fdim*3).half().cuda()
        qkv = qkv.view(bs, sq, 3, nheads, -1)

        mask = None
        out1 = flash_attn(qkv, bs, sq, nheads, scale, mask=mask)
        out2 = std_attn_forward(qkv, bs, sq, nheads, scale, mask=mask)
        print('nheads', nheads, 'same with mask off:', torch.allclose(out1, out2, atol=1e-3))

        mask = (torch.arange(sq) < sq//2).unsqueeze(0).repeat(bs, 1).cuda()
        out1 = flash_attn(qkv, bs, sq, nheads, scale, mask=mask)
        out2 = std_attn_forward(qkv, bs, sq, nheads, scale, mask=mask)

        fmask = mask.float().unsqueeze(2)
        out1 = out1 * fmask
        out2 = out2 * fmask
        print('nheads', nheads, 'same with mask on:', torch.allclose(out1, out2, atol=1e-3))
```

I get 

```
nheads 8 same with mask off: False
nheads 8 same with mask on: False
nheads 16 same with mask off: True
nheads 16 same with mask on: True
```

So the masking seems to work correctly with some number of heads, but the attention does not seem to work with just 8 heads. A head dimension of 1024/8 = 128 seems to be supported here https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attention.py#L88 ?

**Comment by tridao** [comment]
For both cases the two outputs are close (you can print things out to see), `atol=1e-3` is probably too strict (fp16 simply doesn't have that much precision).

We have [tests](https://github.com/HazyResearch/flash-attention/blob/2ed471ecc450801cbe30be2169936a7fe69058c4/tests/test_flash_attn.py#L520) that compare the numerical error of FlashAttention and a reference implementation, and the numerical error is similar to the Pytorch implementation.

You can compare the reference implementation in fp16 and the reference implementation in fp32 (as we did for that test) to see that the numerical error is sometimes more than 1e-3.

These numerical errors are because we're using fp16.
